### PR TITLE
Resource sync fixes: org and team edition.

### PR DESCRIPTION
**Pull request by newswangerd**
## Description
Fixes for:
- AAP-38123: Update sync task to try and handle conflicts.
- AAP-32044: Gracefully handle duplicate name issues for orgs and teams during login

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Steps to Test
1. Create an org with a random ansible ID in Gateway using the name "Conflict"
2. Create an org in any aservice with a randome ansible ID using the name "Conflict"
3. Add a user to the org in Gateway
4. Attempt to login to the service

### Expected Results

You should be able to login no problem, but you won't be marked as a member of the organization until the sync runs.

### Required Actions

- [ ] Requires documentation updates
  
- [ ] Requires downstream repository changes
  
- [ ] Requires infrastructure/deployment changes
  
- [ ] Requires coordination with other teams
  
- [ ] Blocked by PR/MR: #XXX

### CODE REVIEW

File: ansible_base/jwt_consumer/common/auth.py [hunk 1]
```diff
@@ -296,7 +296,12 @@ def process_rbac_permissions(self):
 
             for index in object_indexes:
                 object_data = self.token['objects'][object_type][index]
-                resource, obj = self.get_or_create_resource(object_type, object_data)
+                try:
+                    resource, obj = self.get_or_create_resource(object_type, object_data)
+                except IntegrityError:
+                    logger.warning(f"Got integrity error on {object_data}. Skipping {object_type} assignment.")
```
Comment (john-westcott-iv) [inline ansible_base/jwt_consumer/common/auth.py]: Should we also split out what the `IntegrityError` actually is or will it always be something like `duplicate key`?

Comment (john-westcott-iv) [inline ansible_base/jwt_consumer/common/auth.py]: Also, if I saw this I'd be like why is this happening, how do I fix this? Should we add something like "This will be resolved in the next periodic sync" or something along those lines?

File: ansible_base/jwt_consumer/hub/auth.py [hunk 2]
```diff
@@ -47,7 +48,11 @@ def process_permissions(self):
                     try:
                         team = Resource.objects.get(ansible_id=ansible_id).content_object
                     except Resource.DoesNotExist:
-                        team = self.common_auth.get_or_create_resource('team', team_data)[1]
+                        try:
+                            team = self.common_auth.get_or_create_resource('team', team_data)[1]
+                        except IntegrityError:
+                            logger.warning(f"Got integrity error on {team_data}. Skipping team assignment.")
```
Comment (john-westcott-iv) [inline ansible_base/jwt_consumer/hub/auth.py]: Same comment on this log about should we dump the error and/or give the user more details on how to resolve.

File: ansible_base/resource_registry/models/resource.py [hunk 3]
```diff
@@ -40,6 +41,21 @@ def can_be_managed(self):
     def get_resource_config(self):
         return self.resource_registry.get_config_for_model(model=ContentType.objects.get_for_id(self.content_type_id).model_class())
 
+    def get_conflicting_resource(self, resource_data):
+        filter = {}
+
+        serialized_data = self.serializer_class(data=resource_data)
+        serialized_data.is_valid(raise_exception=True)
+        serialized_data = serialized_data.validated_data
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/models/resource.py]: Do we ever have to consider a case in here where the data isn't valid like the ones with the user IDs and wrap this in a try/catch to prevent an exception?

File: ansible_base/resource_registry/models/resource.py [hunk 4]
```diff
@@ -40,6 +41,21 @@ def can_be_managed(self):
     def get_resource_config(self):
         return self.resource_registry.get_config_for_model(model=ContentType.objects.get_for_id(self.content_type_id).model_class())
 
+    def get_conflicting_resource(self, resource_data):
+        filter = {}
+
+        serialized_data = self.serializer_class(data=resource_data)
+        serialized_data.is_valid(raise_exception=True)
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/models/resource.py]: Should we rename these to `serializer` or something else instead of reassigning to it below for clarity?

File: ansible_base/resource_registry/models/resource.py [hunk 5]
```diff
@@ -40,6 +41,21 @@ def can_be_managed(self):
     def get_resource_config(self):
         return self.resource_registry.get_config_for_model(model=ContentType.objects.get_for_id(self.content_type_id).model_class())
 
+    def get_conflicting_resource(self, resource_data):
+        filter = {}
+
+        serialized_data = self.serializer_class(data=resource_data)
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/models/resource.py]: Do you need to check `can_be_manged` to confirm there is a serializer here? Or does something else do that before calling this method?

File: ansible_base/resource_registry/models/resource.py [hunk 6]
```diff
@@ -40,6 +41,21 @@ def can_be_managed(self):
...
+        serialized_data.is_valid(raise_exception=True)
+        serialized_data = serialized_data.validated_data
+
+        for field in self.serializer_class.UNIQUE_FIELDS:
+            filter[field] = serialized_data[field]
+
+        try:
+            return Resource.get_resource_for_object(self.content_type.get_object_for_this_type(**filter))
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/models/resource.py]: Any possibility that we could get a different type of exception here? Or if we do raising is ok?

File: ansible_base/resource_registry/tasks/sync.py [hunk 7]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
+
+    # If the resource does exist, lets update it first. Hopefully this desn't also result
+    # in a duplicate key error. If it does, we're cooked.
+    elif resp.status_code == 200:
+        data = resp.json()
+        conflict_resource.update_resource(
+            resource_data=data["resource_data"], service_id=data["service_id"], is_partially_migrated=data["is_partially_migrated"]
+        )
+    else:
+        raise ResourceSyncHTTPError
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py]: Should we give some details in this exception as I imagine it would end up getting logged somewhere?

File: ansible_base/resource_registry/tasks/sync.py [hunk 8]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
+    # If the conflicting resource doesn't exist on the server, just go ahead and delete it.
+    # This will most likely happen with resources that are ignored by get_orphan_resources.
+    if resp.status_code == 404:
+        delete_resource(conflict_resource)
+
+    # If the resource does exist, lets update it first. Hopefully this desn't also result
+    # in a duplicate key error. If it does, we're cooked.
+    elif resp.status_code == 200:
+        data = resp.json()
+        conflict_resource.update_resource(
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py:179]: Do we want to try/catch here or otherwise validate that we succeeded?

File: ansible_base/resource_registry/tasks/sync.py [hunk 9]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
     resource_data: dict,
+    api_client: ResourceAPIClient,
     **kwargs,
 ) -> SyncResult:
     """Try to update existing resource."""
     try:
         resource.update_resource(resource_data, partial=True, **kwargs)
     except IntegrityError:  # pragma: no cover
-        return SyncResult(SyncStatus.CONFLICT, manifest_item)
+        # This typically means that there was a duplicate key error. To mitigate this
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py:197]: Is there any chance it wouldn't be a duplicate key and we could up trying to _handle_conflict for something else?

File: ansible_base/resource_registry/tasks/sync.py [hunk 10]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
+        # we will attempt to hanlde the conflicting resource and perform the operation
+        # again.
+        try:
+            _handle_conflict(resource_data, resource.resource_type_obj, api_client)
+            resource.update_resource(resource_data, partial=True, **kwargs)
+        except (ResourceDeletionError, IntegrityError) as e:
+            logger.error(f"Failed to gracefully handle conflict for {resource_data}. Got error {e}.")
+            return SyncResult(SyncStatus.CONFLICT, manifest_item)
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py:205]: Would this def be a CONFLICT and not potentally an an actual ERROR?

File: ansible_base/resource_registry/tasks/sync.py [hunk 11]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
     except Error as e:
+        # Something happened with the database. We don't know what it is. Instead of failing the whole
+        # sync, we'll raise an error and skip this for now.
         logger.error(f"Failed to update resource {resource.ansible_id}. Received error: {e}")
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py]: Can we alter this message to say something like "Will try again on the nest sync"?

File: ansible_base/resource_registry/tasks/sync.py [hunk 12]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
+    resource_service_id: str,
+    api_client: ResourceAPIClient,
+) -> SyncResult:
+    try:
+        Resource.create_resource(
+            resource_type=resource_type,
+            resource_data=resource_data,
+            ansible_id=manifest_item.ansible_id,
+            service_id=resource_service_id,
+        )
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py:228]: The pattern of try and try again is duplicated. It would be really sweet if we could figure out a way of abstracting this so we could work on the "try again" logic in one place instead of two.

File: ansible_base/resource_registry/tasks/sync.py [hunk 13]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
+            _handle_conflict(resource_data, resource_type, api_client)
+            Resource.create_resource(
+                resource_type=resource_type,
+                resource_data=resource_data,
+                ansible_id=manifest_item.ansible_id,
+                service_id=resource_service_id,
+            )
+        except (ResourceDeletionError, IntegrityError) as e:
+            logger.error(f"Failed to gracefully handle conflict for {resource_data}. Got error {e}.")
+            return SyncResult(SyncStatus.CONFLICT, manifest_item)
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py:243]: Same comments about updating the exception with "Will try again" or whatever and also, is this def guaranteed a conflict here?

File: ansible_base/resource_registry/tasks/sync.py [hunk 14]
```diff
@@ -148,22 +148,105 @@ def get_resource_type_names() -> list[str]:
...
+    except Error as e:
+        # Something happened with the database. We don't know what it is. Instead of failing the whole
+        # sync, we'll raise an error and skip this for now.
+        logger.error(f"Failed to update create {manifest_item.ansible_id}. Received error: {e}")
```
Comment (john-westcott-iv) [inline ansible_base/resource_registry/tasks/sync.py]: `update create` is that correct?

File: ansible_base/jwt_consumer/common/auth.py (hunk 1 above)
Comment (newswangerd) [reply ansible_base/jwt_consumer/common/auth.py]: I think it will always be duplicate key, but I added the error message to the log just in case.

File: ansible_base/resource_registry/models/resource.py (hunk 3 above)
Comment (newswangerd) [reply ansible_base/resource_registry/models/resource.py]: I think I prefer to let the caller catch this one. I'm not sure what we'd do if it was not valid. We'd just have to raise another exception.

File: ansible_base/resource_registry/models/resource.py (hunk 6 above)
Comment (newswangerd) [reply ansible_base/resource_registry/models/resource.py]: Potentially "too many objects returned", but if that's the case, then the model is misconfigured and I'd rather raise it.

File: ansible_base/resource_registry/tasks/sync.py (hunk 8 above)
Comment (newswangerd) [reply ansible_base/resource_registry/tasks/sync.py:179]: If this does raise an integrity error, or any other DB related error it will get caught in `_attempt_create_resource` or `_attempt_update_resource`.

File: ansible_base/resource_registry/tasks/sync.py (hunk 10 above)
Comment (newswangerd) [reply ansible_base/resource_registry/tasks/sync.py:205]: Maybe... It's an error, but it is caused by a conflict. I think I'd leave it at conflict.

File: ansible_base/resource_registry/tasks/sync.py (hunk 12 above)
Comment (newswangerd) [reply ansible_base/resource_registry/tasks/sync.py:228]: I thought about making this function recursive, but that got complicated very quickly. This repeats itself, but it's much more understandable.

File: ansible_base/resource_registry/tasks/sync.py (hunk 9 above)
Comment (newswangerd) [reply ansible_base/resource_registry/tasks/sync.py:197]: I'm not sure. I don't know how to tell.

File: ansible_base/resource_registry/tasks/sync.py [hunk 15]
```diff
@@ -148,22 +149,105 @@ def get_resource_type_names() -> list[str]:
...
+    #   incredibly difficult to pull off.
+
+    conflict_resource = resource_type.get_conflicting_resource(resource_data)
+
+    if conflict_resource is None:
+        return
+
+    resp = api_client.get_resource(conflict_resource.ansible_id)
+
+    # If the conflicting resource doesn't exist on the server, just go ahead and delete it.
```
Comment (tznamena) [inline ansible_base/resource_registry/tasks/sync.py:170]: How are we deleting a resource that does not exist?

Comment (newswangerd) [reply ansible_base/resource_registry/tasks/sync.py:170]: It doesn't exist on the resource server, but it does exist on the service where the task is running.

Comment (bhavenst) [reply ansible_base/resource_registry/tasks/sync.py:170]: If you just delete it and sync over a new version, might there be loss of assigned roles?

**Review by bhavenst — APPROVED** [review]
I don't see any glaring issues and John already gave lots of feedback.

**Review by tznamena — APPROVED** [review]
